pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,329 @@
+                            handler_(handler),
+                            master_pid_(master_pid),
+                            babysitter_pid_(babysitter_pid) {
+}
+
+bool IsProcessRunning(pid_t pid) {
+  // "If sig is 0, then no signal is sent, but error checking is still
+  //  performed; this can be used to check for the existence of a process ID
+  //  or process group ID."
+  return kill(pid, 0) == 0;
+}
+
+void ControllerManager::ProcessDeathWatcherThread::Run() {
+  handler_->Message(
+      kInfo, "Watching master (%d) and babysitter (%d) to exit if they do.",
+      master_pid_, babysitter_pid_);
+  while (IsProcessRunning(master_pid_) &&
+         IsProcessRunning(babysitter_pid_)) {
+    timer_->SleepUs(30 * 1000 /* 0.03s */);
+  }
+
+  if (!IsProcessRunning(master_pid_)) {
+    handler_->Message(
+        kInfo, "Master process (%d) exited; controller shutting down.",
+        master_pid_);
+  } else {
+    handler_->Message(
+        kInfo, "Babysitter process (%d) exited; controller shutting down.",
+        babysitter_pid_);
+  }
+
+  // TODO(jefftk): Once the controller is doing real work and there's real
+  // cleanup work to do, we can initiate that from here.
+
+  exit(EXIT_SUCCESS);
+
+  // Because the controller is exiting via exit(), the babysitter process will
+  // see that and quit itself instead of restarting it.
+}
+
+pid_t ControllerManager::ReadPidFromFile(
+    const GoogleString& filename) {
+  if (file_system_->Exists(
+          filename.c_str(), handler_).is_true()) {
+    GoogleString pid_buffer;
+    if (file_system_->ReadFile(
+            filename.c_str(), &pid_buffer, handler_)) {
+      GoogleString pid_buffer_trimmed;
+      TrimWhitespace(pid_buffer, &pid_buffer_trimmed);
+
+      int64 pid64;
+      if (StringToInt64(pid_buffer_trimmed, &pid64)) {
+        pid_t pid = static_cast<pid_t>(pid64);
+        return pid;
+      }
+    }
+  }
+  return -1;  // Failed to read pid.
+}
+
+pid_t ControllerManager::MasterPid() {
+  GoogleString pid_filename = factory_->MasterPidFilename();
+
+  // If the file doesn't exist, or contains the PID of something that isn't
+  // running, block until we have a usable PID.  This can happen if the host
+  // webserver is taking a little while to finish daemonizing.
+  for (int i = 1 /* 1 so we don't warn right away */; true; i++) {
+    pid_t pid = ReadPidFromFile(pid_filename);
+    if (pid != -1 && IsProcessRunning(pid)) {
+      return pid;
+    }
+
+    if (i % 100 == 0) {
+      handler_->Message(
+          kWarning,
+          "Controller stuck waiting for master process PID to appear in %s",
+          pid_filename.c_str());
+    }
+
+    // Sleep instead of spinning.
+    timer_->SleepUs(10 * 1000 /* 0.01s */);
+  }
+}
+
+GoogleString ControllerManager::ControllerPidFilename() {
+  SystemRewriteOptions* options = reinterpret_cast<SystemRewriteOptions*>(
+      factory_->default_options());  // Cast is safe, since we're in system.
```

Comment:
It still shouldn't be a reinterpret_cast, but a static_cast (or dynamic_cast, given your check below)